src/server/pps/server/pipeline_controller_test.go

```diff
@@ -0,0 +1,686 @@
+				return nil
+			}
+			return errors.New("change hasn't reflected")
+		}, backoff.NewTestingBackOff())
+	})
+
+}
+
+func TestAutoscalingBasic(t *testing.T) {
+	stateDriver, infraDriver, mockPachd := ppsMasterHandles(t)
+	pipeline := tu.UniqueString(t.Name())
+	mockAutoscaling(mockPachd, 1, 1)
+	stateDriver.upsertPipeline(&pps.PipelineInfo{
+		Pipeline: client.NewPipeline(pipeline),
+		State:    pps.PipelineState_PIPELINE_STARTING,
+		Details: &pps.PipelineInfo_Details{
+			Autoscaling: true,
+			ParallelismSpec: &pps.ParallelismSpec{
+				Constant: 1,
+			},
+		},
+		Version: 1,
+	})
+	validate(t, stateDriver, infraDriver, []pipelineTest{
+		{
+			pipeline: pipeline,
+			assertWhen: []pps.PipelineState{
+				pps.PipelineState_PIPELINE_RUNNING,
+			},
+			expectedStates: []pps.PipelineState{
+				pps.PipelineState_PIPELINE_STARTING,
+				// STANDBY is set consecutively because ppsutil.SetPipelineState
+				// will still update the PipelineInfo even if the desired state
+				// is already set
+				pps.PipelineState_PIPELINE_STANDBY,
+				pps.PipelineState_PIPELINE_STANDBY,
+				pps.PipelineState_PIPELINE_RUNNING,
+				pps.PipelineState_PIPELINE_STANDBY,
+			},
+		},
+	})
+	require.Equal(t, 1, infraDriver.calls[pipeline][mockInfraOp_CREATE])
+}
+
+func TestAutoscalingManyCommits(t *testing.T) {
+	stateDriver, infraDriver, mockPachd := ppsMasterHandles(t)
+	pipeline := tu.UniqueString(t.Name())
+	mockAutoscaling(mockPachd, 1, 100)
+	inspectCount := 0
+	mockPachd.PFS.InspectCommit.Use(func(context.Context, *pfs.InspectCommitRequest) (*pfs.CommitInfo, error) {
+		inspectCount++
+		return &pfs.CommitInfo{}, nil
+	})
+	stateDriver.upsertPipeline(&pps.PipelineInfo{
+		Pipeline: client.NewPipeline(pipeline),
+		State:    pps.PipelineState_PIPELINE_STARTING,
+		Details: &pps.PipelineInfo_Details{
+			Autoscaling: true,
+			ParallelismSpec: &pps.ParallelismSpec{
+				Constant: 1,
+			},
+		},
+		Version: 1,
+	})
+	validate(t, stateDriver, infraDriver, []pipelineTest{
+		{
+			pipeline: pipeline,
+			assertWhen: []pps.PipelineState{
+				pps.PipelineState_PIPELINE_RUNNING,
+			},
+			expectedStates: []pps.PipelineState{
+				pps.PipelineState_PIPELINE_STARTING,
+				pps.PipelineState_PIPELINE_STANDBY,
+				pps.PipelineState_PIPELINE_STANDBY,
+				pps.PipelineState_PIPELINE_RUNNING,
+				pps.PipelineState_PIPELINE_STANDBY,
+			},
+		},
+	})
+	require.Equal(t, 1, infraDriver.calls[pipeline][mockInfraOp_CREATE])
+	// 2 * number of commits, to capture the user + meta repos
+	require.Equal(t, 200, inspectCount)
+}
+
+func TestAutoscalingManyTasks(t *testing.T) {
+	stateDriver, infraDriver, mockPachd := ppsMasterHandles(t)
+	pipeline := tu.UniqueString(t.Name())
+	mockAutoscaling(mockPachd, 100, 1)
+	mockPachd.PFS.InspectCommit.Use(func(context.Context, *pfs.InspectCommitRequest) (*pfs.CommitInfo, error) {
+		// add some delay so that Scaling Up will kick in before the commit is otherwise considered done
+		time.Sleep(time.Second)
```

Comment:
I'd generally prefer to do things like this using channels (or directly checking the condition you're waiting for) if it's not too messy